src/services/item/plugins/invitation/service.ts

```diff
@@ -0,0 +1,350 @@
+    file: MultipartFile,
+    itemMembershipService: ItemMembershipService,
+  ): Promise<{ memberships: ItemMembership[]; invitations: Invitation[] }> {
+    // verify file is CSV
+    verifyCSVFileFormat(file);
+
+    // get parentItem
+    const item = await repositories.itemRepository.get(itemId);
+    await validatePermission(repositories, PermissionLevel.Admin, actor, item);
+
+    // parse CSV file
+    const { rows, header } = await parseCSV(file.file);
+
+    // check that file has data
+    if (rows.length === 0) {
+      throw new NoDataInFile();
+    }
+
+    // check that the email column has been detected
+    if (!header.includes(EMAIL_COLUMN_NAME)) {
+      throw new MissingEmailColumnInCSVError();
+    }
+
+    // check that all rows have an email set
+    if (rows.some((r) => !Boolean(r.email))) {
+      throw new MissingEmailInRowError();
+    }
+
+    return this._createMembershipsAndInvitationsForUserList(
+      actor,
+      repositories,
+      rows,
+      itemId,
+      itemMembershipService,
+    );
+  }
+
+  async handleCSVInvitations(
+    actor: Member,
+    repositories: Repositories,
+    parentId: string,
+    templateId: string,
+    file: MultipartFile,
+    itemMembershipService: ItemMembershipService,
+  ): Promise<
+    {
+      groupName: string;
+      memberships: ItemMembership[];
+      invitations: Invitation[];
+    }[]
+  > {
+    // verify file is CSV
+    verifyCSVFileFormat(file);
+
+    // get parentItem
+    const parentItem = await repositories.itemRepository.get(parentId);
+    await validatePermission(repositories, PermissionLevel.Admin, actor, parentItem);
+
+    // check that the template exists
+    const templateItem = await repositories.itemRepository.get(templateId);
+    if (!templateItem.id) {
+      throw new TemplateItemDoesNotExist();
+    }
+
+    // parse CSV file
+    const { rows, header } = await parseCSV(file.file);
+
+    // check file is not empty
+    if (rows.length === 0) {
+      throw new NoDataInFile();
+    }
+
+    // if the csv file includes a Group column we will create a structure, so the parent item needs to be a folder
+    const hasGrpCol = header.includes(GROUP_COL_NAME);
+    if (!hasGrpCol) {
+      throw new MissingGroupColumnInCSVError();
+    }
+    if (!isItemType(parentItem, ItemType.FOLDER)) {
+      throw new CantCreateStructureInNoFolderItem();
+    }
+
+    // check that all rows have an email set
+    if (rows.some((r) => !Boolean(r.email))) {
+      throw new MissingEmailInRowError();
+    }
+
+    // check that all rows have a group set
+    if (rows.some((r) => !Boolean(r.group_name))) {
+      throw new MissingGroupInRowError();
+    }
+
+    // group unique names from the group column,
+    // and then create the group folders
+    const dataByGroupName = groupby(rows, (r) => r.group_name);
+
+    const res = Array<{
+      groupName: string;
+      memberships: ItemMembership[];
+      invitations: Invitation[];
+    }>();
```

Comment:
```suggestion
    const res:{
      groupName: string;
      memberships: ItemMembership[];
      invitations: Invitation[];
    }[] = []
```
actually not sure which syntax is the best.

Reply:
Would it make sense to use a map here instead?

Reply:
About the typing, I like the annotation with generics on Arrays, more, as it does feel less like a casting. I think both are equivalent. Let me know which one We should use throughout the codebase.

I tried with the map, but given that I already iterate over the groups, I found it hard (or too involved) to convert to a map. Also in the frontend I was planning to iterate over the elements anyways, so the map would not be very useful I think. Let me know If you think we should still use a map.